### ISSUE TITLE
False duplicate operation name error

### ISSUE THREAD

**Issue by ntziolis** [apollographql/apollo-codegen#293]
Reproduce:

- Create a subscription query with <Name>
- create query with <Name> (same as above)
- when running generation the following error appears
  - There can be only one operation named <Name>

Why is this wrong:
- Operation `type` is not taken into consideration, even though it influences naming of generated types
- The actually generated types are called:
  - <Name>Subscription
  - <Name>Query

This makes for very sensible names which enhances readability instead of having to come up with synthetic names. Same goes for Mutations.

Usage:
 - We opted for making sure the model names are always highly descriptive to each use case (think component)
  - To avoid conflicts we generate separate files for each domain

Note:
I'm using typescript, but I believe this is not typescript specific as it is related to key handling which is likely handled in main lib.

**Comment by lewisf** [comment]
Fair point -- this validation is run on a concatenated document of all graphql operations, but we might want to run the validations before concatenation because the GraphQL spec says that operaton names only need to be unique per document.

@martijnwalraven would a change like this have any negative effects on all the other targets?

**Comment by martijnwalraven** [comment]
@lewisf We need type names to be unique across all query documents on iOS/Android, because the generated types are all part of the same module.

Taking operation type into account when checking uniqueness would make sense however, since we always suffix the generated types with `Query` or `Mutation` (and soon `Subscription`).

My only concern is that we also use operation names to show traces in Engine (although with a hash to disambiguate when necessary). So we'll probably still want to recommend unique operation names even across operation types as a matter of best practice, but we don't have to enforce it.

**Comment by ntziolis** [comment]
Thx for the followup.

For us an optional flag to just ignore duplicate name errors when operation types are different would be enough. Right now we end up with lots of different artificial subscription names and its hard to enforce the same naming structure across a larger team.

The rule:
Just name your subscription the same as the query seems to be easy enough to follow. Especially since it makes total sense to the devs as it automatically provides them with good tooling support like grouping in typeahead etc. (so its not some artificial "i like it better this way" format, but actually it aids their work right away)
